File: go/twelve-days/twelve-days.go
package twelve

import "fmt"

const testVersion = 1

const verseCount = 12

var verseItems = []string{"a Partridge in a Pear Tree", "Turtle Doves", "French Hens", "Calling Birds", "Gold Rings", "Geese-a-Laying", "Swans-a-Swimming", "Maids-a-Milking", "Ladies Dancing", "Lords-a-Leaping", "Pipers Piping", "Drummers Drumming"}

var verseDays = []string{"first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth" }
var verseCounts = []string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "eleven", "twelve" }

func Song()(out string) {
  out = ""
  for i := 1; i <= verseCount; i++ {
		out += Verse(i)
    out += "\n"
	}
  return
}

func Verse(day int) string {
  out := fmt.Sprintf("On the %s day of Christmas my true love gave to me", verseDays[day-1])
  buf := ""
  if day == 1 {
    return out + fmt.Sprintf(", %s", verseItems[0]) + "."
  }

  for i:= 0; i < day; i++ {
    if day != 1 && i == 0 {
      buf = fmt.Sprintf(", and %s", verseItems[i]) + buf
    } else {
      buf = fmt.Sprintf(", %s %s", verseCounts[(i)] ,verseItems[i]) + buf
    }
  }
  return out + buf + "."
}

File: elixir/twelve-days/twelve_days.exs
defmodule TwelveDays do

  @variables [["first", "a Partridge in a Pear Tree"],
              ["second", "two Turtle Doves, and a Partridge in a Pear Tree"],
              ["third", "three French Hens, two Turtle Doves, and a Partridge in a Pear Tree"],
              ["fourth", "four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree"],
              ["fifth", "five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree"],
              ["sixth", "six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree"],
              ["seventh", "seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree"],
              ["eighth", "eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree"],
              ["ninth", "nine Ladies Dancing, eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree"],
              ["tenth", "ten Lords-a-Leaping, nine Ladies Dancing, eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree"],
              ["eleventh", "eleven Pipers Piping, ten Lords-a-Leaping, nine Ladies Dancing, eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree"],
              ["twelfth", "twelve Drummers Drumming, eleven Pipers Piping, ten Lords-a-Leaping, nine Ladies Dancing, eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree"]]

  @doc """
  Given a `number`, return the song's verse for that specific day, including
  all gifts for previous days in the same line.
  """
  @spec verse(number :: integer) :: String.t()
  def verse(number) do
    [day, gifts] = Enum.at(@variables, number-1)
    "On the #{day} day of Christmas my true love gave to me, #{gifts}."
  end

  @doc """
  Given a `starting_verse` and an `ending_verse`, return the verses for each
  included day, one per line.
  """
  @spec verses(starting_verse :: integer, ending_verse :: integer) :: String.t()
  def verses(starting_verse, ending_verse) do
    (starting_verse..ending_verse)
      |> Enum.map(&verse(&1))
      |> Enum.join("\n")
  end

  @doc """
  Sing all 12 verses, in order, one verse per line.
  """
  @spec sing():: String.t()
  def sing do
    verses(1,12)
  end
end

